### PR TITLE
Do not block Translog add on file write

### PR DESCRIPTION
**Pull request by Tim-Brooks**
Currently a TranslogWriter add operation is synchronized. This operation
adds the bytes to the file output stream buffer and issues a write
system call if the buffer is filled. This happens every 8KB which means
that we routinely block other add calls on system writes.

This commit modifies the add operation to simply place the operation in
an array list. The array list if flushed when the sync call occurs or
when 1MB is buffered.

### CODE REVIEW

**Comment by tbrooks8** [comment]
This bring backs #62513, but resolves the issues described in this [comment](https://github.com/elastic/elasticsearch/pull/62513#issuecomment-703008024). Additionally, it adds a test that always checks the flush on write logic.

File: server/src/test/java/org/elasticsearch/index/translog/TranslogTests.java [hunk 1]
```diff
@@ -1315,15 +1315,185 @@ public void testTranslogWriter() throws IOException {
...
+                int initialWriteCalls = writeCalls.get();
+                byte[] bytes = new byte[256];
+                writer.add(ReleasableBytesReference.wrap(new BytesArray(bytes)), 1);
+                writer.add(ReleasableBytesReference.wrap(new BytesArray(bytes)), 2);
+                writer.add(ReleasableBytesReference.wrap(new BytesArray(bytes)), 3);
+                assertThat(persistedSeqNos, empty());
+                assertEquals(initialWriteCalls, writeCalls.get());
+
+                // This will fill the buffer and force a flush
+                writer.add(ReleasableBytesReference.wrap(new BytesArray(bytes)), 4);
```
Comment (henningandersen) [inline server/src/test/java/org/elasticsearch/index/translog/TranslogTests.java:1376]: Would be good to also include testing that the `readBytes` path works, for instance by randomly choosing to do readBytes here instead.

**Review by henningandersen — APPROVED** [review]
LGTM

**Review by dnhatn — APPROVED** [review]
LGTM

**Comment by jpountz** [comment]
I added the `release highlight` label given the spectacular speedup on nightly benchmarks, does it sound right @tbrooks8 @henningandersen ? cc @andreidan who is working on release highlights for 7.10

**Comment by tbrooks8** [comment]
Yes it makes sense to me. The gains depend quite a bit on context. Generally the performance gains are greater in high concurrency / lightly mapped scenarios. And less in low concurrency or heavy analysis scenarios (full text search, etc). 

The taxi / geopoint nightlies are where the gains are most visible. They are also kind of visible on the http logs benchmarks. The nightlies use lower concurrency though, so the performance gains _might_ be larger if the benchmarks were run with 1K clients or something on that order.

**Pull request merged**